File: gold/cpp/set.cpp
#include <iostream>
#include <vector>
using namespace std;

class Tree {

public:
    vector<int> parent, depth;
    Tree(int n) : parent(n + 1), depth(n + 1, 0) {
        for (int i = 0; i <= n; ++i) parent[i] = i;
    }

    int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    void unite(int x, int y) {
        int xRoot = find(x);
        int yRoot = find(y);
        if (xRoot != yRoot) {
            if (depth[xRoot] < depth[yRoot]) {
                parent[xRoot] = yRoot;
            } else if (depth[xRoot] > depth[yRoot]) {
                parent[yRoot] = xRoot;
            } else {
                parent[yRoot] = xRoot;
                depth[xRoot]++;
            }
        }
    }

    bool same(int x, int y) {
        return find(x) == find(y);
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;

    Tree t(n);

    for (int i = 0; i < m; ++i) {
        int operation, a, b;
        cin >> operation >> a >> b;

        if (operation == 0) {
            t.unite(a, b);
        } else if (operation == 1) {
            if (t.same(a, b)) {
                cout << "YES\n";
            } else {
                cout << "NO\n";
            }
        }
    }

    return 0;
}

File: gold/python/set.py
import sys

input=sys.stdin.readline

class Tree:
    def __init__(self, n):
        self.parent=[]
        self.depth=[0]*(n+1)
        for i in range(n+1):
            self.parent.append(i)

    def find(self,x):
        if self.parent[x] is not x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def unite(self,x, y):
        xRoot = self.find(x)
        yRoot = self.find(y)
        if (xRoot != yRoot):
            if (self.depth[xRoot] < self.depth[yRoot]):
                self.parent[xRoot] = yRoot
            elif (self.depth[xRoot] > self.depth[yRoot]):
                self.parent[yRoot] = xRoot
            else:
                self.parent[yRoot] = xRoot
                self.depth[xRoot]+=1

    def same(self,x,y):
        if(self.find(x)==self.find(y)): 
            return True
        else :
            return False
            

def main():

    n, m=map(int,input().split())

    t=Tree(n)
    res=[]

    for i in range(m):
        operation, a, b=map(int,input().split())

        if (operation == 0):
            t.unite(a, b)
        elif (operation == 1):
            if (t.same(a, b)):
                res.append("YES")
            else:
                res.append("NO")

    return res

if __name__ == "__main__":
    res = main()
    for msg in res:
        print(msg)